File: Basics/C#/CSharp.Basics/Samples/FunctionDrivenClass/DogAdoptionService.cs
using CSharp.Basics.Samples.Constructors;
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharp.Basics.Samples.FunctionDrivenClass
{
    public class DogAdoptionService
    {
        public void AdoptDog(Dog dog)
        {
            // run some business logic to adopt the dog
        }

        public AdoptableDog AdoptDog(AdoptableDog dog)
        {
            dog.IsAdopted = true;
            return dog;
        }
    }

    public class AdoptableDog
    {
        public string Name { get; set; }
        public bool IsAdopted { get; set; }
        public AdoptableDog(string name)
        {
            Name = name;
        }
    }
}

File: Basics/Kotlin/app/src/main/java/com/suavepirate/kotlinbasics/samples/functionDrivenClass/DogAdoptionService.kt
package com.suavepirate.kotlinbasics.samples.functionDrivenClass

import com.suavepirate.kotlinbasics.samples.dataClass.Dog



data class AdoptableDog(val name: String) {
    var isAdopted: Boolean = false;
}

class DogAdoptionService {
    fun adoptDog(dog: Dog) {
        // run some business logic to adopt the dog
    }

    fun adoptDog(dog: AdoptableDog): AdoptableDog {
        dog.isAdopted = true
        return dog
    }
}

fun testAdoptions() {
    DogAdoptionService().adoptDog(Dog("Ryder", 2, "Mix Breed")) // void / Unit
    DogAdoptionService().adoptDog(AdoptableDog("Ryder")) // AdoptableDog
}